packages/runtime/src/state/makeswift-api-client.ts

```diff
@@ -119,6 +119,11 @@ async function fetchJson<T>(url: string): Promise<T | null> {
   return response.json()
 }
 
+async function fetchResource<T>(url: string): Promise<T | null> {
+  const result = await fetchJson(url)
+  return !('__typename' in result) ? null : result
+}
```

Comment:
Since we don't 404 on missing resources, make sure we don't return an error payload pretending to be a resource. 

Ideally should validate against a Zod schema of the resource type, but we don't have them readily available.

Reply:
Hmm, I do think we 404 on our API endpoints but don't on GraphQL endpoints. Perhaps we want to fix this at the API layer in the `/api/makeswift/` handlers?

I do agree with you that having proper parsing so we don't add garbage in the cache would also be great. I also understand why we might not want to take that on right now.

Reply:
> Hmm, I do think we 404 on our API endpoints but don't on GraphQL endpoints. Perhaps we want to fix this at the API layer in the /api/makeswift/ handlers?

I think that might break the old versions of the runtime. Specifically for localized resources, we are relying on the fact that fetch of non-existent resource will (silently) succeed but return something w/o an ID.

Reply:
We can, of course, version the API layer and have the new runtime fetch from the new endpoints, but that's more than I can chew right now, haha.

Reply:
Never mind, I just read your PR comment, and see that you weren't suggesting to change api.makeswift.com. I don't see a `makeswift` dir in `src/api/`, though: https://github.com/makeswift/makeswift/tree/main/packages/runtime/src/api

Reply:
Well, I seem to have found the problem 🤦🏻‍♂️ 

![CleanShot 2024-10-15 at 10 54 33@2x](https://github.com/user-attachments/assets/4412a55f-8ff9-44c6-8281-f4c90b62898d)

Reply:
We need to do `res.status(404).json(body)`

Reply:
We seem to have a comment there that says this is intentional... I'm not sure why we'd do that, though.

Reply:
I wonder if we should have a single handler for global elements instead of a separate one for localized global elements 🤔

Reply:
@fikrikarim, maybe you remember why you explicitly wanted to handle this case differently?

Reply:
The comment (`We're not returning 404 if it's null because localized global element is nullable.`) seems to imply that returning 404 will result in a different code path for a localized resource than otherwise, but in fact in the code that makes the call to this "endpoint" (AFAIK [this is the only call site](https://github.com/makeswift/makeswift/blob/4203ec3d4d35c9229e3d81669ac1e54729b99a3a/packages/runtime/src/state/makeswift-api-client.ts#L98)), we silently translate 404 back to `null`, so I think it should be okay to not special-case here.

<img width="1028" alt="fetchJSON" src="https://github.com/user-attachments/assets/7981c957-baf1-480a-aa39-e08620ff5c69">

Reply:
I agree

Reply:
Reworked, confirmed that this behaves as we reasoned.